traits/has_traits.py

```diff
@@ -374,6 +396,10 @@ def update_traits_class_dict(class_name, bases, class_dict):
     class_dict : dict
         A dictionary of class members.
     """
+    # Enable name collision check for user defined classes.
+    check_name_collision = (False if class_dict.get("__module__", "")
+                            .startswith("traits.") else True)
```

Comment:
Nit: It would be better to define this `"traits"` package name as a module level constant. Preferably computed from `__name__`. While it is unlikely that the package will change name, it is rather easy to avoid having to make that assumption.

Nit: The fact that we are doing `startswith("traits.")` instead of `split(".", 1)` here means that if the module name is `"traits"` exactly (i.e. without the `"."`), then `check_name_collision` is true. In other words we will be checking for name collisions for any classes defined in `traits/__init__.py`. There are no classes defined there now, and will unlikely have one... but again, another assumption we can avoid making rather cheaply by using `split`.